### PR TITLE
add a presubmit hook

### LINKED ISSUE #19: firebase?
I would like to use this with firebase, but it's tied directly with iron-ajax, I am not using the rest api with firebase. It would be nice to have more control over the form action...

### PR DESCRIPTION
**Pull request by notwaldorf**
Fixes #19. #26, #63 , #54 :tada: 

You can do anything in this hook. You can change the request. You can cancel the event and do your own, custom, request. You can use it to mail me a block of cheese. The world is yours.

### CODE REVIEW

File: iron-form.html [hunk 1]
```diff
@@ -122,9 +146,9 @@
 
     ready: function() {
       // Object that handles the ajax form submission request.
-      this._requestBot = document.createElement('iron-ajax');
-      this._requestBot.addEventListener('response', this._handleFormResponse.bind(this));
-      this._requestBot.addEventListener('error', this._handleFormError.bind(this));
+      this.request = document.createElement('iron-ajax');
```
Comment (cdata) [inline iron-form.html]: `_requestBot` is dead :crying_cat_face: .... Long live `request`! :crown:

File: iron-form.html [hunk 2]
```diff
@@ -146,20 +170,24 @@
...
+        this.request.body = json;
       } else {
-        this._requestBot.params = json;
+        this.request.params = json;
       }
 
-      this._requestBot.generateRequest();
-      this.fire('iron-form-submit', json);
+      // Allow for a presubmit hook
+      var event = this.fire('iron-form-presubmit', {cancelable: true});
```
Comment (cdata) [inline iron-form.html]: :gem:

File: iron-form.html (hunk 1 above)
Comment (notwaldorf) [inline iron-form.html]: omg i thought this was pointing out a typo and I was like "I CAN'T SEE IT"

It's been a day.

File: iron-form.html [hunk 3]
```diff
@@ -146,20 +170,24 @@
...
+      if(!event.defaultPrevented) {
+        this.request.generateRequest();
+        this.fire('iron-form-submit', json);
```
Comment (cdata) [inline iron-form.html]: Is this event documented anywhere?

Comment (cdata) [inline iron-form.html]: (more specifically with a JSDoc annotation, I mean)

File: iron-form.html [hunk 4]
```diff
@@ -286,10 +314,10 @@
       // Checkboxes and radio buttons should only use their value if they're
       // checked. Custom paper-checkbox and paper-radio-button elements
       // don't have a type, but they have the correct role set.
-      if (el.type == 'checkbox' ||
-          el.type == 'radio' ||
-          el.getAttribute('role') == 'checkbox' ||
-          el.getAttribute('role') == 'radio') {
+      if (el.type === 'checkbox' ||
```
Comment (cdata) [inline iron-form.html]: :heart_eyes_cat:

File: iron-form.html (hunk 3 above)
Comment (notwaldorf) [inline iron-form.html]: `iron-form-submit` is on L125, presubmit on L119

**Comment by cdata** [comment]
This is awesome!

I couldn't help but notice that four bugs are fixed, but only one test is added. Perhaps there should be a test for, "I have replaced `request` entirely with something else that isn't `iron-ajax`?" I'm not sure if any others are that important...

**Comment by notwaldorf** [comment]
Sure! I can add that test. All the 4 tests basically want to do (different) things in a pre-submit hook, which is why they're the same

File: iron-form.html (hunk 3 above)
Comment (cdata) [inline iron-form.html]: Ahh, yes, I missed it :sunglasses:

**Comment by cdata** [comment]
LGTM :+1:

**Pull request merged**